src/sage/misc/sageinspect.py

```diff
@@ -1139,6 +1139,11 @@ def _sage_getargspec_cython(source):
                     defaults=('a string', {(1, 2, 3): True}),
                     kwonlyargs=[], kwonlydefaults=None, annotations={})
     """
+    if not isinstance(source, str):
```

Comment:
what do you think about instead of raising a TypeError to use `assert not None`, and then properly fix the cases in the caller that may pass `None`?

For example, further below one finds
```
try:
            source = sage_getsource(obj)
        except TypeError:  # happens for Python builtins
            source = ''
        if source:
            return inspect.FullArgSpec(*_sage_getargspec_cython(source))
```
with a fall-back if `source == None`. This appears to me like a good pattern that one should encourage.

Reply:
I mean _in principle_ it seems like `TypeError` is the morally correct error to raise (instead of `AssertionError`).

But then if previously `TypeError` is raised for Python builtins, maybe we should change _that_ to… `ValueError` or `RuntimeError`? (which one?)

Reply:
`TypeError` is also raised by Python's inspect in case of Python builtins, so we probably want to keep it.

(My reason for the assertion was that there is already one two lines below, and this is a private helper method anyway.)